rfd/000?-database-access.md

```diff
@@ -0,0 +1,512 @@
+the target database which users log into using `tsh db login` command (see
+below for UX).
+
+### Proxy <---> database service
+
+**Mutual TLS.** The connection that is passed from proxy to a database service
+is upgraded to TLS as well in order to be able to pass identity information
+over the reverse tunnel.
+
+In order to do that, proxy will generate an ephemeral certificate signed
+by the user authority and re-encode all identity and routing information
+extracted from the certificate supplied by the client into it.
+
+### Database service <---> database instance
+
+**Mutual TLS (onprem).**  When connecting to an onprem database instance, the
+database service will use a client certificate for authentication, which it
+will generate on the fly according to the database requirements (for example,
+PostgreSQL requires the database user name to be encoded in the certificate
+as common name).
+
+This also means that the database needs to be configured with Teleport's
+certificate authority and (in general case) use server certificate issued by
+Teleport which can be done using `tctl auth sign` command (see below for UX).
+
+**Password auth (RDS/Aurora).** Amazon RDS/Aurora don't support client cert
+authentication. Instead, they support IAM authentication so Teleport database
+service will generate a short-lived authentication token using RDS API that
+will be used as a password:
+
+https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.IAMDBAuth.html
+
+In order to be able to verify RDS/Aurora server certificate, users will need
+to download RDS root certificate and configure the database service to trust
+it:
+
+https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.SSL.html
+
+## Configuration
+
+### Teleleport database service
+
+The following new configuration section is added to the Teleport file config:
+
+```yaml
+# New global key housing the database service configuration.
+db_service:
+  # Enable or disable the database service.
+  enabled: "yes"
+  # List of the database this service is proxying.
+  databases:
+    # Database instance name, used to refer to an instance in CLI like tsh.
+  - name: "postgres-prod"
+    # Optional free-form verbose description of a database instance.
+    description: "Production instance of PostgreSQL 13.0"
+    # Database procotol, only "postgres" is supported initially.
+    protocol: "postgres"
+    # Database connection endpoint, should be reachable from the service.
+    endpoint: "postgres.internal.example.com:5432"
+    # Optional CA certificate path, e.g. for AWS RDS/Aurora.
+    ca_path: "/path/to/root.pem"
+    # AWS specific configuration for RDS/Aurora databases.
+    aws:
+      # Use IAM authentication with RDS/Aurora database.
+      auth: "iam"
+      # Optional AWS region RDS/Aurora database is running in.
+      region: "us-east-1"
+      # Optional AWS access key, will be taken from environment if unspecified.
+      access_key: "XXX"
```

Comment:
we should exposing static keys as a pattern, always assume only dynamically mounted ones, so remove this

Reply:
We've this open issue regarding IAM: We should review again while testing. https://github.com/gravitational/teleport/issues/2781

Reply:
@webvictim they can still provide static credentials mounted in `~/.aws/credentials` and it will work the same as these inline static keys.

Reply:
It doesn't work exactly the same way, that's the issue. It will work OK as long as nothing _else_ on the instance is also reading from `~/.aws/credentials`. This is the order that the AWS SDK for Go (and other SDKs use) from https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/configuring-sdk.html:

1. Environment variables.
2. Shared credentials file.
3. If your application uses an ECS task definition or RunTask API operation, IAM role for tasks.
4. If your application is running on an Amazon EC2 instance, IAM role for Amazon EC2.

So a problem occurs when someone is relying on an EC2 instance role to provide credentials - if they write a `~/.aws/credentials` file for Teleport, then this will also break whatever depends on the EC2 instance role (as it will start using the credentials file)

On the whole I'm not sure if this is as big of a problem as I'm making it out to be, but I'm trying to call attention to these sorts of cases that will become more common as we get more users. I understand that we don't necessarily want to recommend that people use inline static credentials in their Teleport config files, but there is definitely a benefit in a user being able to provide inline static credentials which use the principle of least privilege and are explicitly only scoped for Teleport DB access.

Reply:
One of the potential use-cases I was thinking about (and partly the reason I originally put these here and made them per db) is that theoretically you can (or may want to) configure a single teleport's db service to proxy to multiple rds/aurora instances in different regions or accounts even, which require different credentials.

Though for this use-case we could probably rely on AWS named credential profiles i.e. have a configuration field that would tell teleport which named profile to load credentials from e.g. given credentials file like this:

```
cat ~/.aws/credentials
[default]
aws_access_key_id = XXX
aws_secret_access_key = YYY
[staging]
aws_access_key_id = ZZZ
aws_secret_access_key = QQQ
```

have the following in db service config to tell it to use staging credentials:

```yaml
aws:
  region: "us-west-1"
  profile: "staging"
```

Reply:
I could see the use of the Teleport DB service as a single gateway to all these different databases in different accounts being really valuable. A developer just being able to say "give me a connection to `staging` as `alice`" and Teleport handling how to do that for them while also keeping a full audit log is a killer feature.

Reply:
I have removed access/secret keys from the config for now.